src/lib/rootfs/image/image.c

```diff
@@ -79,9 +84,64 @@ int rootfs_image_init(char *source, char *mount_dir) {
         }
         read_write = 1;
     } else {
-        if ( ( image_fp = fopen(source, "r") ) == NULL ) { // Flawfinder: ignore
-            singularity_message(ERROR, "Could not open image (read only) %s: %s\n", source, strerror(errno));
-            ABORT(255);
+        if ( ( image_fp = fopen(source, "r") ) != NULL ) { // Flawfinder: ignore
+            singularity_message(VERBOSE, "Opened image (read only, without privileges) %s\n", source);
+        } else {
+            singularity_config_rewind();
+            if ( singularity_config_get_bool("allow privileged fopen", 0) == 0 ) {
+                singularity_message(ERROR, "Privileged fopen is not enabled in config\n" );
+                ABORT(255);
+            }
+
+            singularity_priv_escalate();
+            if ( ( image_fp = fopen(source, "r") ) == NULL ) { // Flawfinder: ignore
+	            singularity_message(ERROR, "Could not open image (read only, with privileges) %s: %s\n", source, strerror(errno));
+                ABORT(255);
+            }
+
+            singularity_message(VERBOSE, "Opened image (read only, with privileges) %s\n", source);
+
+            if ( fstat(fileno(image_fp), &image_stat) != 0 ) {
+                singularity_message(ERROR, "Could not obtain stat on image %s: %s\n", source, strerror(errno));
+                ABORT(255);
+            }
+
+            if ( ( file_grp = getgrgid(image_stat.st_gid) ) == NULL ) {
+                singularity_message(ERROR, "Could not obtain gid of image %s: %s\n", source, strerror(errno));
+                ABORT(255);
+            }
+            singularity_priv_drop();
```

Comment:
Please drop privilege around `L101`; it is not needed for `fstat` or `getgrgid`.  As the latter is pretty complex code inside glibc, I'd rather do it unprivileged.